File: app/src/main/java/com/ne1c/gitterclient/Models/MessageModel.java
package com.ne1c.gitterclient.Models;


import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class MessageModel implements Parcelable {
    public String id;
    public String text;
    public String html;
    public String sent;
    public String editedAt;
    public UserModel fromUser;
    public boolean unread;
    public int readBy;
    public List<Urls> urls;
    public ArrayList<Mentions> mentions;
    //public String issues;
    public int v;

    public static class Mentions implements Parcelable {
        public String screenName;
        public String userId;

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeString(this.screenName);
            dest.writeString(this.userId);
        }

        public Mentions() {
        }

        protected Mentions(Parcel in) {
            this.screenName = in.readString();
            this.userId = in.readString();
        }

        public static final Creator<Mentions> CREATOR = new Creator<Mentions>() {
            public Mentions createFromParcel(Parcel source) {
                return new Mentions(source);
            }

            public Mentions[] newArray(int size) {
                return new Mentions[size];
            }
        };
    }

    public static class Urls implements Parcelable {
        public String url;

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeString(this.url);
        }

        public Urls() {
        }

        protected Urls(Parcel in) {
            this.url = in.readString();
        }

        public static final Creator<Urls> CREATOR = new Creator<Urls>() {
            public Urls createFromParcel(Parcel source) {
                return new Urls(source);
            }

            public Urls[] newArray(int size) {
                return new Urls[size];
            }
        };
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.id);
        dest.writeString(this.text);
        dest.writeString(this.html);
        dest.writeString(this.sent);
        dest.writeString(this.editedAt);
        dest.writeParcelable(this.fromUser, 0);
        dest.writeByte(unread ? (byte) 1 : (byte) 0);
        dest.writeInt(this.readBy);
        dest.writeTypedList(urls);
        dest.writeTypedList(mentions);
        dest.writeInt(this.v);
    }

    public MessageModel() {
    }

    protected MessageModel(Parcel in) {
        this.id = in.readString();
        this.text = in.readString();
        this.html = in.readString();
        this.sent = in.readString();
        this.editedAt = in.readString();
        this.fromUser = in.readParcelable(UserModel.class.getClassLoader());
        this.unread = in.readByte() != 0;
        this.readBy = in.readInt();
        this.urls = in.createTypedArrayList(Urls.CREATOR);
        this.mentions = in.createTypedArrayList(Mentions.CREATOR);
        this.v = in.readInt();
    }

    public static final Parcelable.Creator<MessageModel> CREATOR = new Parcelable.Creator<MessageModel>() {
        public MessageModel createFromParcel(Parcel source) {
            return new MessageModel(source);
        }

        public MessageModel[] newArray(int size) {
            return new MessageModel[size];
        }
    };
}

File: app/src/main/kotlin/com/ne1c/gitteroid/models/data/MessageModel.kt
package com.ne1c.gitteroid.models.data

import android.os.Parcel
import android.os.Parcelable
import java.util.*

data class MessageModel(
        var id: String = "",
        var text: String = "",
        var html: String = "",
        var sent: String = "",
        var editedAt: String = "",
        var fromUser: UserModel? = null,
        var unread: Boolean = false,
        var readBy: Int = 0,
        var v: Int = 0,
        var urls: List<Urls> = Collections.emptyList(),
        var mentions: List<Mentions> = Collections.emptyList()) : Parcelable {

    data class Mentions(var screenName: String = "",
                        var userId: String = "") : Parcelable {
        constructor(source: Parcel) : this(source.readString(), source.readString())

        override fun describeContents() = 0

        override fun writeToParcel(dest: Parcel?, flags: Int) {
            dest?.writeString(screenName)
            dest?.writeString(userId)
        }

        companion object {
            @JvmField val CREATOR: Parcelable.Creator<Mentions> = object : Parcelable.Creator<Mentions> {
                override fun createFromParcel(source: Parcel): Mentions = Mentions(source)
                override fun newArray(size: Int): Array<Mentions?> = arrayOfNulls(size)
            }
        }
    }

    data class Urls(var url: String = "") : Parcelable {
        constructor(source: Parcel) : this(source.readString())

        override fun describeContents() = 0

        override fun writeToParcel(dest: Parcel?, flags: Int) {
            dest?.writeString(url)
        }

        companion object {
            @JvmField val CREATOR: Parcelable.Creator<Urls> = object : Parcelable.Creator<Urls> {
                override fun createFromParcel(source: Parcel): Urls = Urls(source)
                override fun newArray(size: Int): Array<Urls?> = arrayOfNulls(size)
            }
        }
    }

    constructor(source: Parcel) : this(source.readString(),
            source.readString(),
            source.readString(),
            source.readString(),
            source.readString(),
            source.readParcelable(UserModel::class.java.classLoader),
            1.equals(source.readInt()),
            source.readInt(),
            source.readInt(),
            source.createTypedArrayList(Urls.CREATOR),
            source.createTypedArrayList(Mentions.CREATOR))

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel?, flags: Int) {
        dest?.writeString(id)
        dest?.writeString(text)
        dest?.writeString(html)
        dest?.writeString(sent)
        dest?.writeString(editedAt)
        dest?.writeParcelable(fromUser, 0)
        dest?.writeInt((if (unread) 1 else 0))
        dest?.writeInt(readBy)
        dest?.writeInt(v)
        dest?.writeTypedList(urls)
        dest?.writeTypedList(mentions)
    }

    companion object {
        @JvmField val CREATOR: Parcelable.Creator<MessageModel> = object : Parcelable.Creator<MessageModel> {
            override fun createFromParcel(source: Parcel): MessageModel = MessageModel(source)
            override fun newArray(size: Int): Array<MessageModel?> = arrayOfNulls(size)
        }
    }
}
